### PR TITLE
Add `preferTruncationOnSpace` option

### LINKED ISSUE #8: Add option to prefer truncating on spaces if possible
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/cli-truncate/issues/8)

See: https://github.com/sindresorhus/fkill-cli/pull/47#issuecomment-375361462

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/cli-truncate/issues/8).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/cli-truncate/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by ammarbinfaisal**
just finding a space within next 3 characters form actual breeaking point. If the string has to be truncated from start then it searches towards left and if the string has to be truncated from end then it searched towards right, so that the length of the truncated string is not more than the `columns` argument passed.

_some changes to the code were done by `xo --fix` (xo asked for them) like doing destructuring._

Fixes #8

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -27,6 +27,12 @@ cliTruncate('unicorn', 4, {position: 'start'});
 cliTruncate('unicorn', 4, {position: 'middle'});
 //=> 'un…n'
 
+cliTruncate('unicorns rainbow dragons', 20, {position: 'start', preferTruncationOnSpace: true})
+//=> '…rainbow dragons'
+
+cliTruncate('unicorns rainbow dragons', 20, {position: 'middle', preferTruncationOnSpace: true})
+//=> 'unicorns…dragons'
```
Comment (sindresorhus) [inline readme.md]: Add these examples to the option docs instead. Currently, too many examples here. Also include examples of the same string but with the option disabled, so users can easily compare.

File: readme.md [hunk 2]
```diff
@@ -73,6 +79,13 @@ Values: `start` `middle` `end`
 
 Position to truncate the string.
 
+##### preferTruncationOnSpace
+
+Type: `boolean`<br>
+Default: `false`
+
+Truncate the string from a whitespace if it is within 3 characters from the actual breaking point
```
Comment (sindresorhus) [inline readme.md]: What's your reasoning for choosing `3`?

File: index.js [hunk 3]
```diff
@@ -33,13 +35,53 @@ module.exports = (input, columns, opts) => {
...
+			return sliceAnsi(input, 0, nearestSpace) + ellipsis;
+		}
+
 		return sliceAnsi(input, 0, columns - 1) + ellipsis;
 	}
 
 	throw new Error(`Expected \`options.position\` to be either \`start\`, \`middle\` or \`end\`, got ${position}`);
 };
+
+function getIndexOfNearestSpace(str, index, searchRight) {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
function getIndexOfNearestSpace(string, index, shouldSearchRight) {
```

File: test.js [hunk 4]
```diff
@@ -19,4 +19,10 @@ test(t => {
 	t.is(m('unicorn', 6, {position: 'start'}), '…icorn');
 	t.is(m('unicorn', 5, {position: 'middle'}), 'un…rn');
 	t.is(m('unicorns', 6, {position: 'middle'}), 'uni…ns');
+	t.is(m('unicorns rainbow dragons', 20, {position: 'start', preferTruncationOnSpace: true}), '…rainbow dragons');
+	t.is(m('unicorns rainbow dragons', 20, {position: 'end', preferTruncationOnSpace: true}), 'unicorns rainbow…');
+	t.is(m('unicorns rainbow dragons', 6, {position: 'start', preferTruncationOnSpace: true}), '…agons');
+	t.is(m('unicorns rainbow dragons', 6, {position: 'end', preferTruncationOnSpace: true}), 'unico…');
+	t.is(m('unicorns rainbow dragons', 6, {position: 'middle', preferTruncationOnSpace: true}), 'uni…ns');
+	t.is(m('unicorns rainbow dragons', 20, {position: 'middle', preferTruncationOnSpace: true}), 'unicorns…dragons');
```
Comment (sindresorhus) [inline test.js]: These should be in a new test named `preferTruncationOnSpace option`. I think you should also include a version of each assertion with the option disabled, so we can easily compare.

File: readme.md (hunk 2 above)
Comment (sindresorhus) [inline readme.md]:
```suggestion
Truncate the string from a whitespace if it is within 3 characters from the actual breaking point.
```

**Comment by sindresorhus** [comment]
You also need to add it to https://github.com/sindresorhus/cli-truncate/blob/master/index.d.ts

File: readme.md [hunk 5]
```diff
@@ -73,6 +79,13 @@ Values: `start` `middle` `end`
 
 Position to truncate the string.
 
+##### preferTruncationOnSpace
+
+Type: `boolean`<br>
+Default: `false`
```
Comment (sindresorhus) [inline readme.md]: Should it be default?

**Comment by sindresorhus** [comment]
@satyarohith Can you help review?

File: index.js [hunk 6]
```diff
@@ -4,10 +4,12 @@ const stringWidth = require('string-width');
...
 module.exports = (input, columns, opts) => {
 	opts = Object.assign({
-		position: 'end'
+		position: 'end',
+		preferTruncationOnSpace: false
 	}, opts);
 
-	const position = opts.position;
+	const {position} = opts;
+	const {preferTruncationOnSpace} = opts;
```
Comment (satyarohith) [inline index.js]: You can also destructure multiple values. Like this: `const {position, preferTruncationOnSpace} = opts;`

File: readme.md (hunk 2 above)
Comment (ammarbinfaisal) [reply readme.md]: nothing specific; just thought that it would be appropriate. Should we ask the user for it?

File: readme.md (hunk 5 above)
Comment (ammarbinfaisal) [reply readme.md]: don't think so. What do you think?

File: readme.md [hunk 7]
```diff
@@ -73,13 +89,16 @@ Values: `start` `middle` `end`
...
 Position to truncate the string.
 
-##### space
+##### preferTruncationOnSpace
 
 Type: `boolean`<br>
 Default: `false`
 
+Truncate the string from a whitespace if it is within 3 characters from the actual breaking point.
 Add a space between the text and the ellipsis.
```
Comment (sindresorhus) [inline readme.md]: Incorrect merge conflict resolution.

File: readme.md [hunk 8]
```diff
@@ -27,6 +27,22 @@ cliTruncate('unicorn', 4, {position: 'start'});
 cliTruncate('unicorn', 4, {position: 'middle'});
 //=> 'un…n'
 
+cliTruncate('unicorns rainbow dragons', 20, {position: 'start', preferTruncationOnSpace: true})
```
Comment (sindresorhus) [inline readme.md]: These examples should be added at the `preferTruncationOnSpace` option docs. Already too many examples here.

Comment (sindresorhus) [reply readme.md]: And would be more useful to have one before and after example so users can compare what the options does in different scenarios.

File: index.d.ts [hunk 9]
```diff
@@ -7,6 +7,33 @@ declare namespace cliTruncate {
...
+		// Truncate Unicode surrogate pairs
+		cliTruncate('uni\uD83C\uDE00corn', 5);
+		//=> 'uni\uD83C\uDE00…'
+
+		// Truncate fullwidth characters
+		cliTruncate('안녕하세요', 3);
+		//=> '안…'
+		````
+		*/
+		readonly preferTruncationOnSpace?: true | false;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		readonly preferTruncationOnSpace?: boolean;
```

File: index.d.ts [hunk 10]
```diff
@@ -7,6 +7,33 @@ declare namespace cliTruncate {
...
+		//=> '\u001B[31muni\u001B[39m…'
+
+		// Truncate Unicode surrogate pairs
+		cliTruncate('uni\uD83C\uDE00corn', 5);
+		//=> 'uni\uD83C\uDE00…'
+
+		// Truncate fullwidth characters
+		cliTruncate('안녕하세요', 3);
+		//=> '안…'
+		````
```
Comment (sindresorhus) [inline index.d.ts]: Typo

File: index.d.ts [hunk 11]
```diff
@@ -7,6 +7,33 @@ declare namespace cliTruncate {
 		*/
 		readonly position?: 'start' | 'middle' | 'end';
 
+		/**
+		Truncate at different positions
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		Truncate at different positions.
```

File: index.d.ts [hunk 12]
```diff
@@ -7,6 +7,33 @@ declare namespace cliTruncate {
...
+		cliTruncate('\u001B[31municorn\u001B[39m', 4);
+		//=> '\u001B[31muni\u001B[39m…'
+
+		// Truncate Unicode surrogate pairs
+		cliTruncate('uni\uD83C\uDE00corn', 5);
+		//=> 'uni\uD83C\uDE00…'
+
+		// Truncate fullwidth characters
+		cliTruncate('안녕하세요', 3);
+		//=> '안…'
```
Comment (sindresorhus) [inline index.d.ts]: These are just incorrectly copy-pasted.

File: test.js [hunk 13]
```diff
@@ -21,6 +21,15 @@ test('main', t => {
 	t.is(cliTruncate('unicorns', 6, {position: 'middle'}), 'uni…ns');
 });
 
+test('preferTruncationOnSpace option', t => {
+	t.is(cliTruncate('unicorns rainbow dragons', 20, {position: 'start', preferTruncationOnSpace: true}), '…rainbow dragons');
+	t.is(cliTruncate('unicorns rainbow dragons', 20, {position: 'end', preferTruncationOnSpace: true}), 'unicorns rainbow…');
+	t.is(cliTruncate('unicorns rainbow dragons', 6, {position: 'start', preferTruncationOnSpace: true}), '…agons');
+	t.is(cliTruncate('unicorns rainbow dragons', 6, {position: 'end', preferTruncationOnSpace: true}), 'unico…');
+	t.is(cliTruncate('unicorns rainbow dragons', 6, {position: 'middle', preferTruncationOnSpace: true}), 'uni…ns');
+	t.is(cliTruncate('unicorns rainbow dragons', 20, {position: 'middle', preferTruncationOnSpace: true}), 'unicorns…dragons');
```
Comment (sindresorhus) [inline test.js]: Try with some different kinds of strings. Look at my feedback in https://github.com/sindresorhus/cli-truncate/pull/10.

File: index.d.ts (hunk 9 above)
Comment (ammarbinfaisal) [reply index.d.ts]: i don't know typescript 😅  .

**Comment by ammarbinfaisal** [comment]
@sindresorhus is it fine now?

File: index.d.ts [hunk 14]
```diff
@@ -7,6 +7,25 @@ declare namespace cliTruncate {
...
+		//=> '…rainbow dragons'
+
+		cliTruncate('unicorns rainbow dragons', 20, {position: 'middle', preferTruncationOnSpace: true});
+		//=> 'unicorns…dragons'
+
+		cliTruncate('unicorns rainbow dragons', 6, {position: 'end', preferTruncationOnSpace: true});
+		//=> 'unico…'
+		````
+		*/
+		readonly preferTruncationOnSpace?: boolean;
```
Comment (sindresorhus) [inline index.d.ts]: I think it should be moved to after the `space` option. That applies to all the other places it's mentioned too, like the readme.

Comment (ammarbinfaisal) [reply index.d.ts]: okay

File: index.d.ts [hunk 15]
```diff
@@ -28,6 +28,25 @@ declare namespace cliTruncate {
 		```
 		*/
 		readonly space?: boolean;
+
+		/**
+		Truncate at different positions.
```
Comment (sindresorhus) [inline index.d.ts]: This needs the same description as the readme.

File: readme.md [hunk 16]
```diff
@@ -75,9 +78,6 @@ Position to truncate the string.
 
 ##### space
 
-Type: `boolean`<br>
-Default: `false`
-
```
Comment (sindresorhus) [inline readme.md]: ?

Comment (ammarbinfaisal) [reply readme.md]: After checking vs code's git lens I think there was a merge conflict and I used those lines by mistake for preferTruncationOnSpace. I am sorry for messing up PRs so much. I guess I need more experience.

File: index.d.ts (hunk 15 above)
Comment (ammarbinfaisal) [reply index.d.ts]: Done

**Pull request merged**